instrumentation/google-http-client-1.19/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/googlehttpclient/GoogleHttpRequestInstrumentation.java

```diff
@@ -49,78 +50,108 @@ public void transform(TypeTransformer transformer) {
+      }
+      return new AdviceScope(context, context.makeCurrent());
+    }
+
+    public void storeContextToVirtualField(HttpRequest request) {
+      virtualField.set(request, context);
+    }
+  }
+
   @SuppressWarnings("unused")
   public static class ExecuteAdvice {
 
+    @Nullable
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void methodEnter(
-        @Advice.This HttpRequest request,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
+    public static AdviceScope methodEnter(@Advice.This HttpRequest request) {
 
-      context = VirtualField.find(HttpRequest.class, Context.class).get(request);
-      if (context != null) {
+      AdviceScope scope = AdviceScope.fromVirtualFieldContext(request);
+      if (scope != null) {
         // span was created by GoogleHttpClientAsyncAdvice instrumentation below
         // (executeAsync ends up calling execute from a separate thread)
         // so make it current and end it in method exit
-        scope = context.makeCurrent();
-        return;
+        return scope;
       }
-      Context parentContext = currentContext();
-      if (!instrumenter().shouldStart(parentContext, request)) {
-        return;
-      }
-      context = instrumenter().start(parentContext, request);
-      scope = context.makeCurrent();
+
+      return AdviceScope.start(request);
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void methodExit(
         @Advice.This HttpRequest request,
         @Advice.Return HttpResponse response,
         @Advice.Thrown Throwable throwable,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      if (scope == null) {
-        return;
-      }
+        @Advice.Enter @Nullable AdviceScope scope) {
 
-      scope.close();
-      instrumenter().end(context, request, response, throwable);
+      if (scope != null) {
+        scope.end(request, response, throwable);
+      }
     }
   }
 
   @SuppressWarnings("unused")
   public static class ExecuteAsyncAdvice {
 
+    @Nullable
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void methodEnter(
-        @Advice.This HttpRequest request,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-      Context parentContext = currentContext();
-      if (!instrumenter().shouldStart(parentContext, request)) {
-        return;
-      }
-      context = instrumenter().start(parentContext, request);
-      scope = context.makeCurrent();
+    public static AdviceScope methodEnter(@Advice.This HttpRequest request) {
 
-      VirtualField.find(HttpRequest.class, Context.class).set(request, context);
+      AdviceScope locals = AdviceScope.start(request);
+      if (locals != null) {
+        locals.storeContextToVirtualField(request);
+      }
+      return locals;
```

Comment:
Nit, forgot to rename:
```suggestion
      AdviceScope scope = AdviceScope.start(request);
      if (scope != null) {
        scope.storeContextToVirtualField(request);
      }
      return scope;
```